File: android/app/src/main/kotlin/com/player/zen_video_player/InMobiBridge.kt
package com.player.zen_video_player

import android.app.Activity
import android.util.Log
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import com.inmobi.ads.AdMetaInfo
import com.inmobi.ads.InMobiAdRequestStatus
import com.inmobi.ads.InMobiInterstitial
import com.inmobi.ads.listeners.InterstitialAdEventListener
import com.inmobi.sdk.InMobiSdk
import com.inmobi.sdk.SdkInitializationListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

/**
 * Method-channel bridge between the Dart [InMobiAdapter] and the
 * native InMobi Android SDK.
 *
 * Channel name: `zen.ads/inmobi`
 *
 * Methods:
 *   init             { accountId } -> Bool
 *   loadRewarded     { placementId } -> Bool   (true once load succeeded)
 *   showRewarded     () -> Map { shown, rewarded }
 *
 * Note: InMobi exposes rewarded ads through [InMobiInterstitial] on
 * Android — the type name is historical, the actual ad format is set
 * in the InMobi dashboard. This bridge only loads/shows rewarded.
 */
class InMobiBridge(
    private val activity: Activity,
    messenger: BinaryMessenger,
) {
    companion object {
        private const val TAG = "InMobiBridge"
        private const val CHANNEL = "zen.ads/inmobi"
    }

    private val channel = MethodChannel(messenger, CHANNEL)

    private var initialized = false

    private var rewarded: InMobiInterstitial? = null
    private var rewardedReady = false
    private var pendingShowRewarded: MethodChannel.Result? = null
    private var lastRewardEarned = false

    init {
        channel.setMethodCallHandler { call, result -> handle(call, result) }
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "init" -> handleInit(call.argument<String>("accountId").orEmpty(), result)
            "loadRewarded" -> handleLoad(
                call.argument<String>("placementId").orEmpty(),
                result = result,
            )
            "showRewarded" -> handleShow(result)
            else -> result.notImplemented()
        }
    }

    private fun handleInit(accountId: String, result: MethodChannel.Result) {
        if (initialized) {
            result.success(true)
            return
        }
        if (!isValidId(accountId)) {
            Log.w(TAG, "init called with invalid accountId: $accountId")
            result.success(false)
            return
        }
        // Print the GAID once so the developer can copy it into the
        // InMobi (and Unity) dashboard's Test Devices list. Most
        // demand-side platforms only fill ads when test mode is set
        // to "Device" AND the device's advertising ID is registered.
        logAdvertisingId()
        // Verbose SDK logs surface the actual auction-server reason
        // behind INTERNAL_ERROR (which is a deliberate catch-all in
        // the public API). Filter for tag `InMobi` in logcat to see
        // them. Switch to NONE for production builds.
        InMobiSdk.setLogLevel(InMobiSdk.LogLevel.DEBUG)
        Log.w(
            TAG,
            "init: sdkVersion=${InMobiSdk.getVersion()} " +
                "packageName=${activity.packageName} accountId=$accountId",
        )
        // Minimal consent payload. Replace with a proper CMP-driven
        // consent string when targeting GDPR regions.
        val consent = JSONObject().apply {
            put(InMobiSdk.IM_GDPR_CONSENT_AVAILABLE, "true")
            put("gdpr", "0")
        }
        try {
            InMobiSdk.init(activity, accountId, consent, object : SdkInitializationListener {
                override fun onInitializationComplete(error: Error?) {
                    if (error != null) {
                        Log.e(TAG, "init failed: ${error.message}")
                        result.success(false)
                    } else {
                        initialized = true
                        result.success(true)
                    }
                }
            })
        } catch (e: Throwable) {
            Log.e(TAG, "init threw", e)
            result.success(false)
        }
    }

    /**
     * Fetches the device's Google Advertising ID off the main thread
     * (required by [AdvertisingIdClient]) and logs it once at WARN
     * level so it shows up clearly in `adb logcat`. Logged tag and
     * format are stable so a single grep reveals it:
     *
     *     adb logcat -s InMobiBridge | grep GAID
     */
    private fun logAdvertisingId() {
        Thread {
            try {
                val info = AdvertisingIdClient.getAdvertisingIdInfo(activity.applicationContext)
                val id = info.id ?: "<unavailable>"
                val limited = info.isLimitAdTrackingEnabled
                Log.w(
                    TAG,
                    "GAID=$id limitAdTracking=$limited — register this " +
                        "ID under InMobi → Test Devices and Unity → " +
                        "Test Devices to receive test ads.",
                )
            } catch (e: Throwable) {
                Log.w(TAG, "GAID lookup failed: ${e.message}")
            }
        }.apply {
            name = "GAIDLookup"
            isDaemon = true
        }.start()
    }

    private fun handleLoad(
        placementId: String,
        result: MethodChannel.Result,
    ) {
        if (!initialized) {
            Log.w(TAG, "load(rewarded) skipped: SDK not initialized")
            result.success(false)
            return
        }
        val placement = placementId.toLongOrNull()
        if (placement == null || !isValidId(placementId)) {
            Log.w(TAG, "load(rewarded) skipped: invalid placementId=\"$placementId\"")
            result.success(false)
            return
        }

        val resolved = ResultOnce(result)

        val listener = object : InterstitialAdEventListener() {
            override fun onAdLoadSucceeded(ad: InMobiInterstitial, info: AdMetaInfo) {
                rewardedReady = true
                resolved.success(true)
            }

            override fun onAdLoadFailed(ad: InMobiInterstitial, status: InMobiAdRequestStatus) {
                // INTERNAL_ERROR from InMobi is a catch-all. The most common
                // causes are: placement not provisioned for this account,
                // placement type mismatch (e.g. dashboard says Banner but we
                // requested it as a rewarded ad), or device GAID not
                // registered as a Test Device while the account is in
                // test-only mode.
                Log.w(
                    TAG,
                    "load(rewarded) failed: code=${status.statusCode} " +
                        "msg=\"${status.message}\" placementId=$placement",
                )
                resolved.success(false)
            }

            override fun onAdDismissed(ad: InMobiInterstitial) {
                pendingShowRewarded?.success(
                    mapOf("shown" to true, "rewarded" to lastRewardEarned),
                )
                pendingShowRewarded = null
                lastRewardEarned = false
                rewardedReady = false
            }

            override fun onAdDisplayFailed(ad: InMobiInterstitial) {
                pendingShowRewarded?.success(
                    mapOf("shown" to false, "rewarded" to false),
                )
                pendingShowRewarded = null
                lastRewardEarned = false
                rewardedReady = false
            }

            override fun onRewardsUnlocked(
                ad: InMobiInterstitial,
                rewards: Map<Any, Any>?,
            ) {
                lastRewardEarned = true
            }
        }

        try {
            val ad = InMobiInterstitial(activity, placement, listener)
            rewarded = ad
            ad.load()
        } catch (e: Throwable) {
            Log.e(TAG, "load threw", e)
            resolved.success(false)
        }
    }

    private fun handleShow(result: MethodChannel.Result) {
        val ad = rewarded
        if (ad == null || !rewardedReady) {
            result.success(mapOf("shown" to false, "rewarded" to false))
            return
        }
        // Stash the pending result; the listener resolves it from
        // onAdDismissed / onAdDisplayFailed.
        pendingShowRewarded = result
        try {
            ad.show()
        } catch (e: Throwable) {
            Log.e(TAG, "show threw", e)
            pendingShowRewarded?.success(mapOf("shown" to false, "rewarded" to false))
            pendingShowRewarded = null
        }
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        rewarded = null
        pendingShowRewarded = null
    }

    /** Reject obvious placeholder strings so a misconfigured app fails fast. */
    private fun isValidId(value: String): Boolean {
        if (value.isBlank()) return false
        if (value.startsWith("INMOBI_")) return false
        return true
    }

    /** Guarantees a [MethodChannel.Result] is completed at most once. */
    private class ResultOnce(private val inner: MethodChannel.Result) {
        private var completed = false
        fun success(value: Any?) {
            if (completed) return
            completed = true
            inner.success(value)
        }
    }
}

File: ios/Runner/InMobiBridge.swift
import Flutter
import InMobiSDK
import UIKit

/// Method-channel bridge between the Dart `InMobiAdapter` and the
/// native InMobi iOS SDK. Mirrors `InMobiBridge.kt` on Android.
///
/// Channel name: `zen.ads/inmobi`
///
/// Note: InMobi exposes rewarded ads through `IMInterstitial` on iOS
/// — the type name is historical, the actual ad format is set in the
/// InMobi dashboard. This bridge only loads/shows rewarded.
final class InMobiBridge: NSObject {
    private static let channelName = "zen.ads/inmobi"

    private let channel: FlutterMethodChannel

    private var initialized = false

    private var rewarded: IMInterstitial?
    private var rewardedReady = false
    private var pendingShowRewarded: FlutterResult?
    private var rewardedDelegate: InterstitialEventForwarder?
    private var lastRewardEarned = false

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: InMobiBridge.channelName,
            binaryMessenger: messenger
        )
        super.init()
        channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
    }

    func dispose() {
        channel.setMethodCallHandler(nil)
        rewarded = nil
        rewardedDelegate = nil
        pendingShowRewarded = nil
    }

    // MARK: - Dispatch

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        switch call.method {
        case "init":
            handleInit(accountId: args?["accountId"] as? String ?? "", result: result)
        case "loadRewarded":
            handleLoad(
                placementId: args?["placementId"] as? String ?? "",
                result: result
            )
        case "showRewarded":
            handleShow(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Init

    private func handleInit(accountId: String, result: @escaping FlutterResult) {
        if initialized {
            result(true)
            return
        }
        guard isValidId(accountId) else {
            NSLog("[InMobiBridge] init called with invalid accountId: \(accountId)")
            result(false)
            return
        }
        // Minimal consent payload. Replace with a proper CMP-driven
        // consent string when targeting GDPR regions.
        let consent: [String: Any] = [
            IM_GDPR_CONSENT_AVAILABLE: "true",
            "gdpr": "0"
        ]
        IMSdk.initWithAccountID(accountId, consentDictionary: consent) { [weak self] error in
            if let error = error {
                NSLog("[InMobiBridge] init failed: \(error.localizedDescription)")
                result(false)
            } else {
                self?.initialized = true
                result(true)
            }
        }
    }

    // MARK: - Load

    private func handleLoad(
        placementId: String,
        result: @escaping FlutterResult
    ) {
        if !initialized {
            result(false)
            return
        }
        guard let placementInt = Int64(placementId), isValidId(placementId) else {
            result(false)
            return
        }

        let resolved = ResultOnce(result)
        let forwarder = InterstitialEventForwarder(
            onLoaded: { [weak self] in
                self?.rewardedReady = true
                resolved.success(true)
            },
            onLoadFailed: { error in
                NSLog("[InMobiBridge] load failed: \(error?.localizedDescription ?? "unknown")")
                resolved.success(false)
            },
            onDismissed: { [weak self] in
                guard let self = self else { return }
                self.pendingShowRewarded?([
                    "shown": true,
                    "rewarded": self.lastRewardEarned
                ])
                self.pendingShowRewarded = nil
                self.lastRewardEarned = false
                self.rewardedReady = false
            },
            onDisplayFailed: { [weak self] in
                guard let self = self else { return }
                self.pendingShowRewarded?([
                    "shown": false,
                    "rewarded": false
                ])
                self.pendingShowRewarded = nil
                self.lastRewardEarned = false
                self.rewardedReady = false
            },
            onRewardsUnlocked: { [weak self] in
                self?.lastRewardEarned = true
            }
        )

        let ad = IMInterstitial(placementId: placementInt)
        ad.delegate = forwarder
        rewarded = ad
        rewardedDelegate = forwarder

        ad.load()
    }

    // MARK: - Show

    private func handleShow(result: @escaping FlutterResult) {
        guard let ad = rewarded, rewardedReady else {
            result(["shown": false, "rewarded": false])
            return
        }
        pendingShowRewarded = result
        guard let root = topViewController() else {
            pendingShowRewarded = nil
            result(["shown": false, "rewarded": false])
            return
        }
        ad.show(from: root)
    }

    private func topViewController() -> UIViewController? {
        guard let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let root = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController
        else { return nil }
        var top = root
        while let presented = top.presentedViewController { top = presented }
        return top
    }

    private func isValidId(_ value: String) -> Bool {
        if value.isEmpty { return false }
        if value.hasPrefix("INMOBI_") { return false }
        return true
    }
}

// MARK: - Helpers

/// Forwards InMobi rewarded events (delivered through the
/// `IMInterstitialDelegate` API) to closures so the bridge stays small.
private final class InterstitialEventForwarder: NSObject, IMInterstitialDelegate {
    private let onLoaded: () -> Void
    private let onLoadFailed: (Error?) -> Void
    private let onDismissed: () -> Void
    private let onDisplayFailed: () -> Void
    private let onRewardsUnlocked: () -> Void

    init(
        onLoaded: @escaping () -> Void,
        onLoadFailed: @escaping (Error?) -> Void,
        onDismissed: @escaping () -> Void,
        onDisplayFailed: @escaping () -> Void,
        onRewardsUnlocked: @escaping () -> Void
    ) {
        self.onLoaded = onLoaded
        self.onLoadFailed = onLoadFailed
        self.onDismissed = onDismissed
        self.onDisplayFailed = onDisplayFailed
        self.onRewardsUnlocked = onRewardsUnlocked
    }

    func interstitialDidFinishLoading(_ interstitial: IMInterstitial) {
        onLoaded()
    }

    func interstitial(
        _ interstitial: IMInterstitial,
        didFailToLoadWithError error: IMRequestStatus
    ) {
        onLoadFailed(error)
    }

    func interstitial(
        _ interstitial: IMInterstitial,
        didFailToPresentWithError error: IMRequestStatus
    ) {
        onDisplayFailed()
    }

    func interstitialDidDismiss(_ interstitial: IMInterstitial) {
        onDismissed()
    }

    func interstitial(
        _ interstitial: IMInterstitial,
        rewardActionCompletedWithRewards rewards: [AnyHashable: Any]
    ) {
        onRewardsUnlocked()
    }
}

/// Guarantees a [FlutterResult] is invoked at most once.
private final class ResultOnce {
    private var inner: FlutterResult?
    init(_ result: @escaping FlutterResult) {
        self.inner = result
    }
    func success(_ value: Any?) {
        guard let inner = inner else { return }
        self.inner = nil
        inner(value)
    }
}
